protoc-gen-kotlin/protoc-gen-kotlin-common/src/main/kotlin/pbandk/gen/CodeGenerator.kt

```diff
@@ -27,15 +27,19 @@ open class CodeGenerator(val file: File, val kotlinTypeMappings: Map<String, Str
     }
 
     protected fun writeEnumType(type: File.Type.Enum) {
-        // Enums are data classes w/ a single value and a companion object with known values
-        line().line("data class ${type.kotlinTypeName}(override val value: Int) : pbandk.Message.Enum {").indented {
-            line("companion object : pbandk.Message.Enum.Companion<${type.kotlinTypeName}> {").indented {
-                type.values.forEach { line("val ${it.kotlinValueName} = ${type.kotlinTypeName}(${it.number})") }
-                line()
-                line("override fun fromValue(value: Int) = when (value) {").indented {
-                    type.values.forEach { line("${it.number} -> ${it.kotlinValueName}") }
-                    line("else -> ${type.kotlinTypeName}(value)")
-                }.line("}")
+        // Enums are data classes w/ a value and a name, and a companion object with known values
+        line().line("sealed class ${type.kotlinTypeName}(override val value: Int, override val name: String) : pbandk.Message.NamedEnum {").indented {
+            line("override fun equals(other: kotlin.Any?) = other is ${type.kotlinTypeName} && other.value == value")
+            line("override fun hashCode() = value.hashCode()")
+            line("override fun toString() = \"${type.kotlinTypeName}.\$name(value=\$value)\"")
+            line()
+            type.values.forEach { line("object ${it.kotlinValueClassName} : ${type.kotlinTypeName}(${it.number}, \"${it.kotlinValueStringName}\")") }
```

Comment:
🔧 I think you'd want to use `it.name` instead of `it.kotlinValueStringName` for the `name` property. I assume the intent is to eventually use the enum's `name` property when parsing from JSON, and for that we want the original enum name in the `.proto` definition, not pb-and-k's mangled version of it.

Reply:
Yeah, that seems correct.  Was going to move into JSON soon, but will switch this to using the `name` for now.

Reply:
Wait a sec... don't we want to use `GENERATE_IF_EMPTY`, and not `PASSWORD_GRANT_GENERATE_IF_EMPTY` for json?  If not, do we want to support both?

Reply:
Nevermind, for canonical json, we'll need to be using `PASSWORD_GRANT_GENERATE_IF_EMPTY`, as you know.  This means it is what will also be stored in the database unfortunately, but not the end of the world.